### DISCUSSION TITLE
Floating timestamp according to visible messages in list

### DISCUSSION THREAD

**Discussion by itayganor** [petyosi/react-virtuoso#334] [Q&A]
Hi!
Do you know that feature in WhatsApp (or Telegram or other IM platforms) where there's a blue sticky bubble that shows the date of the conversation? And when you scroll past a bubble, the sticky one changes its value?

![image](https://user-images.githubusercontent.com/6430313/114191990-a8e85380-9955-11eb-94f1-2df4b6324c5e.png)  ![image](https://user-images.githubusercontent.com/6430313/114192009-b00f6180-9955-11eb-8228-f945bce9b9f7.png)

(screenshots are stolen from [this thread](https://stackoverflow.com/questions/39397127/how-to-display-a-date-like-whatsapp-on-scrolling-listview))

That's exactly what I'm trying to implement (except each message updates the bubble, instead of only when the day has changed).

### Implementation methods

I was thinking about using `VirtuosoGroup` but it's not a group, just a floating bubble that shows the current date.

Then I thought maybe I could utilize `rangeChanged` to know which message index is currently being rendered at the top - but since I have to use `overscan`, and [`rangeChanged` includes `overscan`](https://virtuoso.dev/range-change-callback/), I can't.

Following @petyosi's demo on #118 I tried to use `IntersectionObserver` to know when a message goes out of the screen. The given demo in that issue isn't exactly working, So with help from [this thread](https://stackoverflow.com/a/51976805/6340965), I made this:

https://codesandbox.io/s/messages-list-intersection-observers-ifjc3?file=/src/MessageList/Message.tsx

![9dc97d85-2922-4a1a-81ff-37fb9720f7ca](https://user-images.githubusercontent.com/6430313/114192912-b05c2c80-9956-11eb-8f2d-5a15bc344df1.gif)

As you can see it's *almost* working- but doesn't work very well if you scroll too fast, and its behavior is really weird when scrolling up.

---

It feels like listening to scroll events on the scrollable container is the only choice, but it's difficult to calculate the current topmost visible message because the rows are windowed by `Virtuoso`.

Do you have an idea of a possible way to implement this wonder?

Thanks! <3

**Comment by petyosi** [comment]
Why not groups? If you make the group transparent and center the contents it should work like WhatsApp. It will also have the scrolling dividers stick to the header.

**Reply by itayganor** [reply]
Groups could fit but I need the scroll position anyway to scroll-sync another chat screen (on the other half of the screen) according to the current position.

**Reply by petyosi** [reply]
I need to make reverse scrolling and item prepending properly working in grouped mode...

**Reply by isBatak** [reply]
@petyosi do you have any update on item prepending in grouped mode?
Currently, I'm getting this error `Error: Virtuoso: prepending items does not work with groups`.
